debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/NonStreamingWal2JsonMessageDecoder.java

```diff
@@ -58,10 +58,17 @@ public void processMessage(ByteBuffer buffer, ReplicationMessageProcessor proces
             final Instant commitTime = Conversions.toInstant(dateTime.systemTimestamp(timestamp));
             final Array changes = message.getArray("change");
 
-            Iterator<Entry> it = changes.iterator();
-            while (it.hasNext()) {
-                Value value = it.next().getValue();
-                processor.process(new Wal2JsonReplicationMessage(txId, commitTime, value.asDocument(), containsMetadata, !it.hasNext(), typeRegistry));
+            // WAL2JSON may send empty changes that still have a txid. These events are from things like vacuum,
+            // materialized view, DDL, etc. They still need to be processed for the heartbeat to fire.
+            if(changes.isEmpty()) {
+                processor.process(null);
+            }
+            else {
+                Iterator<Entry> it = changes.iterator();
```

Comment:
Is it possible to change it to for-each loop?

Reply:
I started to do that, but the logic uses the hasNext() to determine if that change is the last change in a transaction.